i.eodag: add i.eodag module

HamedElgizery opened the pull request:
Adding **i.eodag** addon module. The module can be used for downloading different imagery datasets e.g. Landsat, Sentinel, and MODIS. **EODAG** [1] provides a **single API interface** to request imagery datasets from a number of different providers.

[1] https://eodag.readthedocs.io/en/stable/index.html

HamedElgizery committed:
i.eodag: create basic structure

HamedElgizery committed:
i.eodag: set default values for start and end

HamedElgizery committed:
i.eodag: use current compuataionl for searching by default

HamedElgizery committed:
i.eodag: add download of the search results to directory 'output'

HamedElgizery committed:
i.eodag: add downloading by IDs

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+
+
+# %Module
+# % description: Eodag interface to install imagery datasets from various providers.
+# % keyword: eodag
```
```suggestion
# % description: Eodag interface to install imagery datasets from various providers.
# % keyword: imagery
# % keyword: eodag
```
First keyword should be module category.

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+# %option
+# % key: start
+# % type: string
+# % description: Start date ('YYYY-MM-DD'), by default it is 60 days ago
```
Date should be any ISO formated string and also include time information (not only date)...

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+# %option
+# % key: end
+# % type: string
+# % description: End date ('YYYY-MM-DD')
```
As with start...

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+
+
+# %flag
+# % key: l
```
We may consider a format option (with standard parser option G_OPT_M_FORMAT) for all printing (CSV, JSON, plain, ...)

ninsbl replied:
Handled further here: https://github.com/HamedElgizery/grass-addons/issues/4

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+
+# %flag
+# % key: l
+# % description: List the search result without downloading
```
There will be probably several types of information to print. It may make sense to have here options instead of a flag....

lucadelu replied:
Yes, @ninsbl do you have an output of i.sentinel.download to show to @HamedElgizery ?

ninsbl replied:
Further below an output from i.sentinel.downoad.

What I think could be different information that should be possible to print to user are:
- current eodag configuration (and file location)
- configuration of a chosen provider (e.g. when print option and provider option are given)
- available providers
- available providers for a chosen product (if only a product is chosen)
- available products
- available products of a chosen provider (e.g. when print option and provider option are given)
- queryables / metadata of product types (requires both provider and product option).

So a `print` option may make sense with possible answers: "configuration,provider,product,product_metadata"

There may be more. But printable data has different complexity (can be a simple list where plain output would be sufficient or a hirarchical structure (e.g. product metadata) where JSON output would be more appropriate. But I would say printing is a next step when basic query and download functionality is more mature... Thus also a format option for printing...
```
"29bbec5a-9ab4-41e5-88f5-d716242d944f S3B_SL_1_RBT____20240526T195554_20240526T195854_20240528T010235_0179_093_242_1080_PS2_O_NT_004 2024-05-26T19:55:54Z cloudcover_NA SL_1_RBT___ 417.02 MB\nf702eab4-1644-447b-8046-34e09b5dd250 S3B_SL_1_RBT____20240526T181455_20240526T181755_20240528T010629_0179_093_241_1080_PS2_O_NT_004 2024-05-26T18:14:54Z cloudcover_NA SL_1_RBT___ 425.36 MB\naed9bd3f-e134-45ef-8b33-576523312296 S3B_SL_1_RBT____20240526T195254_20240526T195554_20240528T010205_0179_093_242_0900_PS2_O_NT_004 2024-05-26T19:52:54Z cloudcover_NA SL_1_RBT___ 404.04 MB\nad03b03b-44e8-41d8-abb9-6713c833d325 S3B_SL_1_RBT____20240526T213353_20240526T213653_20240528T010044_0179_093_243_0900_PS2_O_NT_004 2024-05-26T21:33:53Z cloudcover_NA SL_1_RBT___ 391.18 MB\n8d55c4fc-662c-411b-87ba-3b3029826e0d S3A_SL_1_RBT____20240526T171249_20240526T171549_20240528T031123_0179_112_383_1080_PS1_O_NT_004 2024-05-26T17:12:48Z cloudcover_NA SL_1_RBT___ 424.19 MB\n227018c9-87f7-4bef-b564-40257ab7a4c2 S3A_SL_1_RBT____20240526T171549_20240526T171849_20240528T031123_0179_112_383_1260_PS1_O_NT_004 2024-05-26T17:15:48Z cloudcover_NA SL_1_RBT___ 425.10 MB\n96674eba-7e6c-4020-ace0-f4c8ee355467 S3A_SL_1_RBT____20240526T185348_20240526T185648_20240528T044547_0179_112_384_1080_PS1_O_NT_004 2024-05-26T18:53:47Z cloudcover_NA SL_1_RBT___ 425.44 MB\n7b52e1cb-4bca-46fb-8642-224166705c46 S3A_SL_1_RBT____20240526T203147_20240526T203447_20240528T062503_0179_112_385_0900_PS1_O_NT_004 2024-05-26T20:31:47Z cloudcover_NA SL_1_RBT___ 405.00 MB\n6392a06d-ba7a-4a7d-aad3-b2732edcc21f S3A_SL_1_RBT____20240526T203447_20240526T203747_20240528T062504_0179_112_385_1080_PS1_O_NT_004 2024-05-26T20:34:47Z cloudcover_NA SL_1_RBT___ 406.67 MB\nfa9475ae-aebb-4760-a674-bba5335da40a S3B_SL_1_RBT____20240527T093548_20240527T093848_20240528T102505_0179_093_250_1800_PS2_O_NT_004 2024-05-27T09:35:48Z cloudcover_NA SL_1_RBT___ 448.52 MB\n0a900a66-9b06-4a16-aadf-c85eba792814 S3B_SL_1_RBT____20240527T111347_20240527T111647_20240528T102940_0180_093_251_1620_PS2_O_NT_004 2024-05-27T11:13:47Z cloudcover_NA SL_1_RBT___ 425.28 MB\n86361c46-41af-4553-bf31-3ceaf541634f S3B_SL_1_RBT____20240527T111947_20240527T112247_20240528T103038_0179_093_251_1980_PS2_O_NT_004 2024-05-27T11:19:47Z cloudcover_NA SL_1_RBT___ 432.68 MB\nffd76101-4ff0-45bb-bfe4-b47175b1a61e S3B_SL_1_RBT____20240527T111647_20240527T111947_20240528T103009_0179_093_251_1800_PS2_O_NT_004 2024-05-27T11:16:47Z cloudcover_NA SL_1_RBT___ 422.23 MB\nba940202-dd53-4e7c-908e-a7cf0b5b79d2 S3B_SL_1_RBT____20240527T192644_20240527T192944_20240528T103921_0180_093_256_0900_PS2_O_NT_004 2024-05-27T19:26:43Z cloudcover_NA SL_1_RBT___ 400.10 MB\n99240ed7-0449-4b91-bca1-07160b7e50a0 S3B_SL_1_RBT____20240527T192944_20240527T193244_20240528T103942_0179_093_256_1080_PS2_O_NT_004 2024-05-27T19:29:43Z cloudcover_NA SL_1_RBT___ 424.03 MB\nb7b763c6-6a3b-4376-87a1-478c22031720 S3B_SL_1_RBT____20240527T175144_20240527T175444_20240528T104434_0179_093_255_1260_PS2_O_NT_004 2024-05-27T17:51:44Z cloudcover_NA SL_1_RBT___ 417.55 MB\nd72f5d69-ddc7-447f-893c-3bb95531a8ef S3B_SL_1_RBT____20240527T093848_20240527T094148_20240528T102533_0179_093_250_1980_PS2_O_NT_004 2024-05-27T09:38:48Z cloudcover_NA SL_1_RBT___ 442.27 MB\nf50eed12-42ad-4c01-8790-20c6aa8fc8e8 S3B_SL_1_RBT____20240527T174844_20240527T175144_20240528T104415_0179_093_255_1080_PS2_O_NT_004 2024-05-27T17:48:44Z cloudcover_NA SL_1_RBT___ 425.20 MB\n174ae551-1b7e-4168-aac8-5af4df6dd5ff S3B_SL_1_RBT____20240527T211043_20240527T211343_20240528T144646_0179_093_257_1080_PS2_O_NT_004 2024-05-27T21:10:42Z cloudcover_NA SL_1_RBT___ 402.48 MB\n874b59da-443d-4ee7-b814-e9c84a3e43c7 S3B_SL_1_RBT____20240527T210743_20240527T211043_20240528T144626_0179_093_257_0900_PS2_O_NT_004 2024-05-27T21:07:42Z cloudcover_NA SL_1_RBT___ 395.54 MB\n300facba-66dc-4fd3-a42a-e4f8d9e29dbb S3A_SL_1_RBT____20240527T083342_20240527T083642_20240528T181029_0180_113_007_1800_PS1_O_NT_004 2024-05-27T08:33:41Z cloudcover_NA SL_1_RBT___ 442.87 MB\n41385ccf-9068-4041-ba2b-64fa6b158feb S3A_SL_1_RBT____20240527T101141_20240527T101441_20240528T195034_0179_113_008_1620_PS1_O_NT_004 2024-05-27T10:11:41Z cloudcover_NA SL_1_RBT___ 429.01 MB\nf44cf156-78d4-45c2-9cde-b5e0301662bc S3A_SL_1_RBT____20240527T101741_20240527T102041_20240528T195035_0179_113_008_1980_PS1_O_NT_004 2024-05-27T10:17:41Z cloudcover_NA SL_1_RBT___ 453.01 MB\n18831c4d-f223-4abf-8248-703056117101 S3A_SL_1_RBT____20240527T101441_20240527T101741_20240528T195035_0179_113_008_1800_PS1_O_NT_004 2024-05-27T10:14:41Z cloudcover_NA SL_1_RBT___ 450.67 MB\n"
```

ninsbl replied:
Handled further here: [here](https://github.com/HamedElgizery/grass-addons/issues/4)

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+        except Exception as e:
+            gs.warning(_("Could not create directory {}").format(dir))
+            return 1
+    else:
```
After returns you do not need else. If you run pylint on the script it will tell you as well....

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+            "lonmax": info["ne_long"],
+            "latmax": info["nw_lat"],
+        }
+    else:
```
Also here "else" and indent  is not needed.

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+
+
+def download_by_id(query_id: str):
+    print(f"Searching for product ending with: {query_id[-min(len(query_id), 8):]}")
```
`gs.verbose`  instead of `print`?

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+
+
+def setup_environment_variables():
+    os.environ[f"EODAG__{options['provider']}__DOWNLOAD__EXTRACT"] = str(flags["e"])
+    os.environ[f"EODAG__{options['provider']}__DOWNLOAD__DELETE_ARCHIV"] = str(
+        flags["d"]
+    )
+
+    if options["output"]:
+        os.environ[f"EODAG__{options['provider']}__DOWNLOAD__OUTPUTS_PREFIX"] = options[
+            "output"
+        ]
```
THis only works if `provider` is required input...

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+        # TODO: Check that the product exists, could be handled by catching exceptions when searching...
+        product_type = options["dataset"]
+
+        # TODO: Allow user to specify a shape file path
```
If one works in projected CRS it may be more convenient to be able to provide tha AOI in e.g. a GeoJSON... In principle any OGR readable format could be OK...

ninsbl replied:
Handled further here: https://github.com/HamedElgizery/grass-addons/issues/3

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+        delta_days = timedelta(60)
+        if not options["start"]:
+            start_date = date.today() - delta_days
+            start_date = start_date.strftime("%Y-%m-%d")
```
datetime.from_isoformat would be more flexible....

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,319 @@
+
+        if end_date < start_date:
+            gs.fatal(
+                _(f"End Date ({end_date}) cannot come before Start Date ({start_date})")
```
Translatable strings should be formated (not f-strings)

ninsbl reviewed (commented):
Some first comments. Very good start, Hamed.

@wenzeslaus Is it OK if we use a draft PR for technical GSoC discussions, Or should we have such discussions at a different place for now?

veroandreo reviewed (commented):
Thanks @HamedElgizery! I haven't tested yet, but some general comments:
- the .gitignore should not be part of the PR
- the html only includes the following sections: Description, Examples, See also, Author, all the rest is added by the parser from the code (the .py file in this case). Check the html file in i.landsat.download or i.sentinel.download for guidance

wenzeslaus commented:
> @wenzeslaus Is it OK if we use a draft PR for technical GSoC discussions, Or should we have such discussions at a different place for now?

Draft PR is appropriate. A great place to have a discussion over specific code!

HamedElgizery committed:
i.eodag: delete .gitignore

HamedElgizery committed:
i.eodag: change strings to translatable strings

HamedElgizery committed:
i.eodag: change time to isoformat

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,326 @@
+            os.makedirs(dir)
+            return 0
+        except Exception as e:
+            gs.warning(_("Could not create directory {}").format(dir))
```
This case should lead to a fatal error (`gs.fatal()`)

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,326 @@
+from grass.exceptions import ParameterError
+
+
+def create_dir(dir):
```
This function could be shortened with either https://docs.python.org/3/library/os.html#os.makedirs
or probably even better https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,326 @@
+    # are we in LatLong location?
+    kv = gs.parse_command("g.proj", flags="j")
+    if "+proj" not in kv:
+        gs.fatal("Unable to get bounding box: unprojected location not supported")
```
```suggestion
        gs.fatal(_("Unable to get bounding box: unprojected location not supported"))
```

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,326 @@
+    }
+
+
+def download_products(products_list):
```
A function with just one line of code just increases the number of lines of code. But it can be OK if you intend to extend it later on...

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,326 @@
+
+def download_by_id(query_id: str):
+    gs.verbose(
+        _(f"Searching for product ending with: {query_id[-min(len(query_id), 8):]}")
```
Use format, not f-string for all translatable strings...
There is some more use of f-strings below...

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,326 @@
+    # products: https://github.com/CS-SI/eodag/blob/develop/eodag/resources/product_types.yml
+
+    # TODO: Add option for setting a differnt config file path
+    global options, flags
```
no need for global, if you put the following line (`options, flags = gs.parser()`) into the `if __name__ == "__main__":` section

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,326 @@
+
+# %option
+# % key: id
+# % type: string
```
Is it possible users would want to download a lager number of files that way? Then we should consider also the possibility of "file" input.

ninsbl replied:
Handled further here: https://github.com/HamedElgizery/grass-addons/issues/6

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,326 @@
+    else:
+
+        items_per_page = 20
+        # TODO: Check that the product exists, could be handled by catching exceptions when searching...
```
See also the code in the GIST I referenced in the Google doc. You may pick elemnts from that for implementing some of the options...

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,326 @@
+        from eodag.utils import get_geometry_from_various
+
+        # for debugging -> 3
+        setup_logging(verbose=2)
```
You can use GRASS debug messages (`gs.debug()`)

lucadelu replied:
I think you have to use gs.debug()

HamedElgizery replied:
Noted, I think I need setup_logging as well to get logs of eodag, e.g. when it is searching? Not sure if there is a way to pipeline the eodag logs into gs.debug...

`setup_logging()` comes from the eodag library to view its logs, that is why I am using it for now.

ninsbl replied:
I see. You can match the logging level to the GRASS verbosity / debug level (https://grass.osgeo.org/grass83/manuals/g.message.html), then.
Maybe possible to use [ProgressCallback](https://eodag.readthedocs.io/en/stable/api_reference/utils.html#callbacks)? It should be available in download functions?

HamedElgizery replied:
> You can match the logging level to the GRASS verbosity / debug level (https://grass.osgeo.org/grass83/manuals/g.message.html), then.

I have just added that in this commit: 610f9c9

ninsbl reviewed (commented):
Nice. It is taking shape. Please run pylint on your code to get some more hints on code formatting...

ninsbl commented:
BTW, if searching does not require credentials, it would be good to add a basic unit-test for searching...

HamedElgizery commented:
> BTW, if searching does not require credentials, it would be good to add a basic unit-test for searching...

Most of the providers doesn't requirer credentials for searching, while others require credentials... So, if the user set a specific provider for searching (e.g. provider=usgs). If usgs requires credentials, then eodag will attempt to search using a different provider, if it didn't find the usgs credentials...

So, I think we could leave it for eodag to handle it, and just send a warning if the credentials are not there for that specified provider.

Something along the lines of: 
"Credentials for {provider} was not found... Searching might be automatically rerouted to a different provider."

What do you think?

ninsbl commented:
>  (...) So, I think we could leave it for eodag to handle it, and just send a warning if the credentials are not there for that specified provider. (...)
> What do you think?

Sounds convinceing. We should then make sure in the module logic that the provider is not a required option (and consider potential consequences, e.g. lines 193, 199, ...)...

HamedElgizery committed:
i.eodag: normalize time to utc in isoformat

HamedElgizery committed:
i.eodag: remove timezone metadata

HamedElgizery commented:
After a second thought, I think making the provider required by the user would constraint the behavior of the library better, and only searching in that provider. By that I mean, preventing eodag from automatically rerouting the searching process to a different provider, and instead it just terminates with a relevant error message or "0 results found", when the searching fails for the given provider.

That will make it more convenient when using it inside the other modules e.g. i.landsat.download & i.sentinel.download.

We can also provide a method for the user to see which providers have the datasets they are looking for... and then they can use one of these to search for the intended product.

Nevertheless, we could add a search feature without setting the "provider" later on.

ninsbl commented:
> After a second thought, I think making the provider required by the user would constraint the behavior of the library better, 

Agreed, Also, by requesting a specific provider I would expect results to be more reliable. Anyway, you may also consider allowing multiple input (comma separated list) in the providers option... (which is probably a solution inbetween).

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -0,0 +1,343 @@
+# % guisection: Output
+# %end
+
+# %option G_OPT_M_DIR
+# % key: format
+# % type: string
+# % description: Don't download and show/save search results as: plain;Plain text output;json;JSON
+# % required: no
+# %end
```
```suggestion
# %option
# % key: format
# % type: string
# % description: Output format
# % required: no
# % options: plain,json
# % answer: json
# %end
```
The appropriate standard parser option is [G_OPT_F_FORMAT](https://grass.osgeo.org/grass84/manuals/parser_standard_options.html). Please note that it is quite new in GRASS GIS 8.4 (actually added by another GSoC student: https://github.com/OSGeo/grass/pull/3704/files). So for the time beeing, (and to support GRASS GIS < 8.4), you may use the above suggestion / work-around...

For printing you probably have to add another `print` option... (as format only defines the format for printing (or output to file)...

HamedElgizery committed:
i.eodag: add option ot choose eodag config file

HamedElgizery committed:
i.eodag: add search with disabled fallback (no_fallback_search)

Used to search in one and only one provider, if it doesn't find the
intended product, it doesn't search in other providers

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -228,10 +234,32 @@ def normalize_time(datetime_str: str):
+            return SearchResult([])
+    except Exception as e:
+        gs.verbose(e)
+        gs.fatal(_("Server error, try again."))
```
The `gs.fatal` causes exit. The return should not be reached. I suggest merging your use of `gs.verbose` and `gs.fatal` to a `gs.warning` (just guessing from your code).

HamedElgizery replied:
Oh yeah right, I should remove the "return" instead, because if I got into the exception, that means I couldn't connect to the server, so it should just terminate with an error message... will fix that.

ninsbl commented on src/imagery/i.eodag/i.eodag.py:
```diff
@@ -308,9 +336,13 @@ def main():
-        gs.message(_(f"Found {num_results} matching scenes " f"of type {product_type}"))
+        # print([p.properties["storageStatus"] for p in search_results])
+
+        gs.message(
```
This should probably be a verbose message (`gs.verbose`)

HamedElgizery committed:
i.eodag: map eodag logging level to grass debug level

ninsbl commented:
I filed some issues here: https://github.com/HamedElgizery/grass-addons/issues so we can keep better track of the larger development items. The discussion about specific code (and more minor adjustments) can continue here.
Anyway, the PR gets already a bit messy with the many comments... 

So, my suggestion is to fix the main blockers for merging (esp. the manual), then you add a warning message at the top of the module that it is experimental and the API is not finalized (so users know everything can change) and then we merge a first version.

That would easy testing and feedback from the wider community...

HamedElgizery committed:
i.eodag: refactor html manual

ninsbl commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -1,90 +1,66 @@
+Search available scenes:
+
+<div class="code"><pre>
+i.eodag.py  start=2022-05-25 end=2022-06-01 \
```
```suggestion
i.eodag  start=2022-05-25 end=2022-06-01 \
```

ninsbl commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -1,90 +1,66 @@
+Download all available scenes:
+
+<div class="code"><pre>
+i.eodag.py  start=2022-05-25 end=2022-06-01 \
```
```suggestion
i.eodag  start=2022-05-25 end=2022-06-01 \
```

ninsbl commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -1,90 +1,66 @@
+Download selected scenes by ID:
+
+<div class="code"><pre>
+i.eodag.py provider=cop_dataspace \
```
```suggestion
i.eodag provider=cop_dataspace \
```

ninsbl reviewed (approved):
Great work!
For me, this is ready to go. Feel free to commit the suggested changes or address them later as you prefer.

Will you merge @lucadelu or @veroandreo

HamedElgizery committed:
Update src/imagery/i.eodag/i.eodag.html

Co-authored-by: Stefan Blumentrath <stefan.blumentrath@gmx.de>

HamedElgizery committed:
Update src/imagery/i.eodag/i.eodag.html

Co-authored-by: Stefan Blumentrath <stefan.blumentrath@gmx.de>

HamedElgizery committed:
Update src/imagery/i.eodag/i.eodag.html

Co-authored-by: Stefan Blumentrath <stefan.blumentrath@gmx.de>

ninsbl committed:
Merge branch 'grass8' into i.eodag

HamedElgizery committed:
Fixed a bug

HamedElgizery committed:
Merge branch 'i.eodag' of github.com:HamedElgizery/grass-addons into i.eodag

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,66 @@
+The module utilizes the <a href="https://eodag.readthedocs.io/en/stable/">EODAG API</a>, as a single interface to search for datasets on the providers that EODAG supports.
+
+<p>
+To be able to download data through i.eodag, you will need to register for the providers you are interested in.
+See <a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/register.html">providers registration</a>, and <a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/configure.html">configure EODAG</a> for more details about configuring the providers' credentials.
```
```suggestion
To be able to download data through <em>i.eodag</em>, users will need to register with the providers they are interested in.
See <a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/register.html" target="_blank">providers registration</a>, and <a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/configure.html" target="_blank">configure EODAG</a> sections for more details about registration and configuration of the providers' credentials.
```

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,66 @@
+<p>
+To be able to download data through i.eodag, you will need to register for the providers you are interested in.
+See <a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/register.html">providers registration</a>, and <a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/configure.html">configure EODAG</a> for more details about configuring the providers' credentials.
+<p>
```
```suggestion

```

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,66 @@
+<p>
+
+<p>
+Currently, only products which footprint intersects current computation
+region extent (area of interest, AOI) are filtered.
```
```suggestion
Currently, only products which footprint intersects the current computational
region extent (area of interest, AOI) are retrieved.
```

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,66 @@
+To download all scenes found within the time frame provided, the user must
+remove the <b>l</b> flag and provide an <b>output</b> directory. Otherwise,
+files will be downloaded into <em>/tmp</em> directory.
+To download only selected scenes, one or more IDs must be provided
+through the <b>id</b> option file and an
+<b>output</b> directory.
```
```suggestion
To download only selected scenes, one or more IDs must be provided
through the <b>id</b> option.
```

veroandreo replied:
Edit this once the <file> option is in place

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,66 @@
+
+<h2>EXAMPLES</h2>
+
+Search available scenes:
+
+<div class="code"><pre>
+i.eodag  start=2022-05-25 end=2022-06-01 \
+    dataset=S2_MSI_L2A provider=cop_dataspace -l
+<pre></div>
```
```suggestion
Search and list the available Sentinel 2 scenes in the Copernicus Data Space Ecosystem:

<div class="code"><pre>
i.eodag -l start=2022-05-25 end=2022-06-01 \
    dataset=S2_MSI_L2A provider=cop_dataspace
<pre></div>
```

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,66 @@
+    dataset=S2_MSI_L2A provider=cop_dataspace -l
+<pre></div>
+
+Download all available scenes:
+
+<div class="code"><pre>
+i.eodag  start=2022-05-25 end=2022-06-01 \
+    dataset=S2_MSI_L2A provider=cop_dataspace
+<pre></div>
```
```suggestion
Download all available scenes in the tmp directory:

<div class="code"><pre>
i.eodag start=2022-05-25 end=2022-06-01 \
    dataset=S2_MSI_L2A provider=cop_dataspace
<pre></div>
```

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,66 @@
+    dataset=S2_MSI_L2A provider=cop_dataspace
+<pre></div>
+
+Download selected scenes by ID:
+
+<div class="code"><pre>
+i.eodag provider=cop_dataspace \
+    id=S2B_MSIL2A_20240526T080609_N0510_R078_T37SDD_20240526T094753,S2B_MSIL2A_20240529T081609_N0510_R121_T37SED_20240529T124818 \
+    -e config=custom/eodag_config.yaml
+<pre></div>
```
```suggestion
Download and extract only selected scenes into the tmp directory:

<div class="code"><pre>
i.eodag -e \
    id=S2B_MSIL2A_20240526T080609_N0510_R078_T37SDD_20240526T094753,S2B_MSIL2A_20240529T081609_N0510_R121_T37SED_20240529T124818 \
    provider=cop_dataspace \
    config=custom/eodag_config.yaml
<pre></div>
```

veroandreo replied:
A note on where to store the config file (or where eodag will look for it, if not provided) might be useful as well as an example of config file.

ninsbl replied:
This seems addressed.

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,66 @@
+<h2>REQUIREMENTS</h2>
+
+<ul>
+    <li><a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/install.html">eodag library</a> (install with <code>pip install eodag</code>)</li>
```
```suggestion
    <li><a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/install.html">EODAG library</a> (install with <code>pip install eodag</code>)</li>
```

veroandreo replied:
Let's agree on how to write EODAG/eodag and keep it consistent through docs

HamedElgizery replied:
EODAG makes more sense to me.

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,66 @@
+    <li><a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/install.html">eodag library</a> (install with <code>pip install eodag</code>)</li>
+</ul>
+
+<h2>SEE ALSO</h2>
```
perhaps we could add i.landsat, i.sentinel, i.modis here?

veroandreo reviewed (commented):
I just left some comments on the html file. I think we can merge after those are solved.

HamedElgizery committed:
Update src/imagery/i.eodag/i.eodag.html

Co-authored-by: Veronica Andreo <veroandreo@gmail.com>

HamedElgizery committed:
Update src/imagery/i.eodag/i.eodag.html

Co-authored-by: Veronica Andreo <veroandreo@gmail.com>

neteler commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,65 @@
+<h2>DESCRIPTION</h2>
+<H3>WARNING: I.EODAG IS STILL UNDER DEVELOPMENT. THIS IS AN EXPERIMENTAL VERSION.</H3>
+
+<em>i.eodag</em> allows to search and download imagery datasets, e.g. Sentinel, Landsat, and MODIS, from a number of different providers.
```
@HamedElgizery Please break long lines in the HTML file as per style guide:

https://github.com/OSGeo/grass/blob/main/doc/development/style_guide.md#tool-documentation
-> More notes on markup:
-->  In order to minimize potential git merge conflicts, please break a line at approximately 70-80 chars.

Thanks!

HamedElgizery replied:
Thank you, should be fixed now.

HamedElgizery commented:
Thanks for the comments, @veroandreo . I will fix them ASAP.

HamedElgizery committed:
Update src/imagery/i.eodag/i.eodag.html

Co-authored-by: Veronica Andreo <veroandreo@gmail.com>

HamedElgizery committed:
Update src/imagery/i.eodag/i.eodag.html

Co-authored-by: Veronica Andreo <veroandreo@gmail.com>

HamedElgizery committed:
Update src/imagery/i.eodag/i.eodag.html

Co-authored-by: Veronica Andreo <veroandreo@gmail.com>

HamedElgizery committed:
Reformat manual

veroandreo commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,113 @@
+<h2>REQUIREMENTS</h2>
+
+<ul>
+    <li><a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/install.html">eodag library</a>
```
```suggestion
    <li><a href="https://eodag.readthedocs.io/en/stable/getting_started_guide/install.html">EODAG library</a>
```

veroandreo committed:
Update src/imagery/i.eodag/i.eodag.html

veroandreo reviewed (approved):
I think this first version is ready to merge

neteler commented on src/imagery/i.eodag/i.eodag.html:
```diff
@@ -0,0 +1,113 @@
+
+<h2>SEE ALSO</h2>
+
+<em>
+<a href="https://grass.osgeo.org/grass83/manuals/addons/i.landsat.html">i.landsat</a>
+</em>
+<p>
+<em>
+<a href="https://grass.osgeo.org/grass83/manuals/addons/i.sentinel.html">i.sentinel</a>
+</em>
+</p>
+<p>
+<em>
+<a href="https://grass.osgeo.org/grass83/manuals/addons/i.modis.html">i.modis</a>
+</em>
+</p>
```
```suggestion
<em>
<a href="i.landsat.html">i.landsat</a>,
<a href="i.sentinel.html">i.sentinel</a>,
<a href="i.modis.html">i.modis</a>
</em>
```

This snippet can be simplified: addon references do not need the full URL.

HamedElgizery committed:
Update src/imagery/i.eodag/i.eodag.html

Co-authored-by: Markus Neteler <markus@neteler.org>

HamedElgizery committed:
Merge branch 'grass8' into i.eodag

ninsbl commented:
Taking the liberty to merge this first version. Further improvements can be discussed in new PRs.

ninsbl commented:
Congrats to your first contributed module, @HamedElgizery !

HamedElgizery commented:
> Congrats to your first contributed module, @HamedElgizery !

Thanks @ninsbl, and thank you all for the help!
